### PR TITLE
V2: training with ensemble

### PR DESCRIPTION
**Pull request by hwpang**
## Description
This PR contains efforts adding ensemble modeling, originally implemented in https://github.com/chemprop/chemprop/pull/608. It depends on https://github.com/chemprop/chemprop/pull/608 and should be merged in after it.

- [x] Add ensemble (different initial seed)
- [x] Add ensemnle to CLI test

## Relevant issues
https://github.com/chemprop/chemprop/issues/581

### CODE REVIEW

File: chemprop/cli/train.py [hunk 1]
```diff
@@ -697,81 +697,86 @@ def main(args):
...
-        model = build_model(args, train_dset)
-        logger.info(model)
+            train_dset, val_dset, test_dset = build_datasets(args, train_data, val_data, test_data)
 
-        monitor_mode = "min" if model.metrics[0].minimize else "max"
-        logger.debug(f"Evaluation metric: '{model.metrics[0].alias}', mode: '{monitor_mode}'")
+            if no_cv:
+                output_dir = args.output_dir / f"model_{model_idx}"
+            else:
+                output_dir = args.output_dir / f"model_{model_idx}" / f"fold_{fold_idx}"
```
Comment (davidegraff) [inline chemprop/cli/train.py:720]:
```suggestion
            if not no_cv:
                output_dir = output_dir / f"fold_{fold_idx}"
```
and add:
```
output_dir = args.output_dir / f"model_{model_idx}"
```
before the inner loop

**Review by davidegraff — COMMENTED** [review]
If I understand the model correctly here, it generalizes _all_ model training into training an ensemble of $j$ models, where $j=1$ by default. So it seems like really the major change to make this work was adding an outer loop to `train.main()` that essentially calls `main()` $j$ times. If it's not too much work, I think that rather than indenting all of the training code, we refactor `main()` to instead call `train_model()` $j$ times, where `train_model()` is largely the same as `main()` is right now, i.e., the outer loop of your current implementation. I think the benefit of a change like this is the new `train_model()` method is a more clear analogy of where a user might go to copy code from if they'd like to implement their own training code using chemprop.

**Comment by hwpang** [comment]
> If I understand the model correctly here, it generalizes _all_ model training into training an ensemble of j models, where j=1 by default. So it seems like really the major change to make this work was adding an outer loop to `train.main()` that essentially calls `main()` j times. If it's not too much work, I think that rather than indenting all of the training code, we refactor `main()` to instead call `train_model()` j times, where `train_model()` is largely the same as `main()` is right now, i.e., the outer loop of your current implementation. I think the benefit of a change like this is the new `train_model()` method is a more clear analogy of where a user might go to copy code from if they'd like to implement their own training code using chemprop.

I love this refactoring idea! I will work on this.

**Comment by KnathanM** [comment]
When ensembling, how do you test the overall performance of the averaged prediction of the models? I see that when training each model, the train/val/test splits are different. Likely all of the data is used for training at least one model, but each individual model has a set of test data set aside to give a fair test metric for the individual model. But does the individual test performance of each model clearly translate to an overall performance of the ensemble of models?

**Comment by KnathanM** [comment]
I see one style inconsistency that you could leave or fix. `model_idx` is used to do two thing: increment the seed `seed = args.seed + model_idx` and to name the directory `output_dir = args.output_dir / f"model_{model_idx}"`. But the seed is set before `train_model` is called while the directory is named as the first thing `train_model` does. You could change it to something like:
```
def main():
    seed = args.seed + model_idx
    output_dir = args.output_dir / f"model_{model_idx}"
    train_model(args, output_dir, no_cv, seed, format_kwargs, featurization_kwargs)
```
or 
```
def train_model():
    seed = args.seed + model_idx
    output_dir = args.output_dir / f"model_{model_idx}"

train_model(args, model_idx, no_cv, format_kwargs, featurization_kwargs)
```
My preference to the latter because it reduces how many parameters `train_model` takes.

**Comment by hwpang** [comment]
> When ensembling, how do you test the overall performance of the averaged prediction of the models? I see that when training each model, the train/val/test splits are different. Likely all of the data is used for training at least one model, but each individual model has a set of test data set aside to give a fair test metric for the individual model. But does the individual test performance of each model clearly translate to an overall performance of the ensemble of models?

Good question. I take a quick look into v1 and it seems that the test set is hold out and test all the ensemble models at once. I will think about how to modify the current code.

**Comment by hwpang** [comment]
@KnathanM Thanks for the review and the comments! I have addressed all of them. I think this is ready for another round of review after the tests pass.

**Review by KnathanM — APPROVED** [review]
